File: main/dijkstrashortreach/dijkstrashortreach.py
#!/usr/bin/env python3
"""Dijkstra: Shortest Reach 2 (HackerRank problem)"""

import collections


class VertexCostPair:
    """A key-value pair where the key is a vertex and the value is a cost."""

    __slots__ = ('_vertex', 'cost')

    def __init__(self, vertex, cost):
        """Creates a pair for a vertex with a specified initial cost."""
        self._vertex = vertex
        self.cost = cost

    def __iter__(self):
        """Yields the vertex and cost values, to allow unpacking."""
        yield self._vertex
        yield self.cost

    @property
    def vertex(self):
        """Retrieves the vertex, which should never change."""
        return self._vertex


class PrimHeap:
    """A binary minheap suitable for Prim's and Dijkstra's algorithms."""

    __slots__ = ('_heap', '_map', '_len')

    def __init__(self, capacity):
        """Creates an empty PrimHeap with the specified maximum capacity."""
        self._heap = [None] * capacity  # index -> (vertex, cost)
        self._map = [None] * capacity   # vertex -> index
        self._len = 0

    def __len__(self):
        """The current size of the heap."""
        return self._len

    def __bool__(self):
        """True if nonempty, false if empty."""
        return self._len != 0

    def push_or_update(self, vertex, cost):
        """
        Decreases a vertex's cost to the specified value unless its cost is
        already lower. If the vertex doesn't exist, adds it with that cost.
        """
        child = self._map[vertex]

        if child is None:
            child = self._len
            self._len += 1
            item = VertexCostPair(vertex, cost)
        else:
            item = self._heap[child]
            if item.cost <= cost:
                return
            item.cost = cost

        self._sift_up(child, item)

    def pop(self):
        """
        Removes and returns the cheapest vertex and its cost from the heap.
        """
        assert self._len, "can't pop from an empty heap"

        result = self._heap[0]
        self._map[result.vertex] = None
        self._len -= 1

        if self._len:
            item = self._heap[self._len]
            self._heap[self._len] = None
            self._sift_down(0, item)
        else:
            self._heap[0] = None

        return result

    def _sift_up(self, child, item):
        """Sifts the specified item up from the specified child index."""
        while child:
            parent = (child - 1) // 2
            if self._heap[parent].cost <= item.cost:
                break
            self._heap[child] = self._heap[parent]
            self._map[self._heap[parent].vertex] = child
            child = parent

        self._heap[child] = item
        self._map[item.vertex] = child

    def _sift_down(self, parent, item):
        """Sifts the specified item down from the specified parent index."""
        while True:
            child = self._pick_child(parent)
            if child is None or item.cost <= self._heap[child].cost:
                break

            self._heap[parent] = self._heap[child]
            self._map[self._heap[child].vertex] = parent
            parent = child

        self._heap[parent] = item
        self._map[item.vertex] = parent

    def _pick_child(self, parent):
        """Picks index of the child that can switch places with the parent."""
        left = parent * 2 + 1
        if left >= self._len:
            return None

        right = left + 1
        if right != self._len:
            if self._heap[right].cost < self._heap[left].cost:
                return right

        return left


OutEdge = collections.namedtuple('OutEdge', ('dest', 'weight'))
OutEdge.__doc__ = """
    A weighted directed edge representation for an adjacency list.
    """


def read_value():
    """Reads a line of input as a single integer."""
    return int(input())


def read_record():
    """Reads a line of input as a sequences of integers."""
    return map(int, input().split())


def read_graph():
    """
    Reads a weighted edge-list representation of an undirected graph, emitting
    an adjacency list.
    """
    vertex_count, edge_count = read_record()
    adj = [[] for _ in range(vertex_count + 1)]  # +1 for 1-based indexing

    for _ in range(edge_count):
        first, second, weight = read_record()
        adj[first].append(OutEdge(second, weight))
        adj[second].append(OutEdge(first, weight))

    return adj


def dijkstra(adj, start):
    """
    Runs Dijkstra's algorithm on the graph represented by the adjacency list
    adj, from the specified start vertex. Returns costs, where costs[i] is the
    minimum edge weight sum from start to i, or None if there is no such path.
    """
    costs = [None] * len(adj)
    heap = PrimHeap(len(adj))
    heap.push_or_update(start, 0)

    while heap:
        src, src_cost = heap.pop()
        costs[src] = src_cost

        for dest, weight in adj[src]:
            if costs[dest] is None:
                heap.push_or_update(dest, src_cost + weight)

    return costs


def print_costs(costs, start):
    """
    Print costs of all but the unused zeroth vertex and the start vertex.
    Vertices that are unreachable have infinite cost, but we print -1 instead.
    """
    print(*((-1 if cost is None else cost)
            for cost in costs[1:start] + costs[(start + 1):]))


def run():
    """Reads input and solves as many test cases as requested."""
    for _ in range(read_value()):
        adj = read_graph()
        start = read_value()
        costs = dijkstra(adj, start)
        print_costs(costs, start)


if __name__ == '__main__':
    run()

File: main/dijkstrashortreach/dijkstrashortreach.rb
#!/usr/bin/env ruby
$VERBOSE = true

# A (key, value) entry in a PrimHeap.
Entry = Struct.new(:key, :value)

# A binary minheap, modified for use in Prim's or Dijkstra's algorithm.
class PrimHeap
  def initialize(min_key, max_key)
    raise ArgumentError, 'negative length range' if min_key > max_key

    # create and populate the key => index map
    @map = (0..(max_key - min_key)).to_a
    min_key.times { @map.unshift(nil) }

    # create and populate the index => (key, value) entries
    @entries = (min_key..max_key).map { |key| Entry.new(key, Float::INFINITY) }
  end

  def empty?
    @entries.empty?
  end

  def pop
    ret = @entries.first
    @map[ret.key] = nil

    if @entries.one?
      @entries.clear
    else
      @entries[0] = @entries.pop
      sift_down(0)
    end

    ret
  end

  def decrease(key, value)
    child = @map[key]
    return if child.nil?

    ent = @entries[child]
    return if ent.value <= value

    ent.value = value
    sift_up(child)
    nil
  end

  private def sift_up(child)
    until child.zero?
      parent = (child - 1) / 2
      break if order_ok?(parent, child)

      swap(child, parent)
      child = parent
    end

    record(child)
  end

  private def sift_down(parent)
    loop do
      child = pick_child(parent)
      break if child.nil? || order_ok?(parent, child)

      swap(parent, child)
      parent = child
    end

    record(parent)
  end

  private def pick_child(parent)
    left = parent * 2 + 1
    return nil if left >= @entries.size

    right = left + 1
    right == @entries.size || order_ok?(left, right) ? left : right
  end

  private def swap(src, dest)
    @entries[dest], @entries[src] = @entries[src], @entries[dest]
    record(src)
  end

  private def record(src)
    @map[@entries[src].key] = src
  end

  private def order_ok?(parent, child)
    @entries[parent].value <= @entries[child].value
  end
end

# The endpoints of an undirected edge, without weight information.
Endpoints = Struct.new(:u, :v) do
  def initialize(u, v)
    if u > v
      super(v, u)
    else
      super(u, v)
    end

    freeze
  end
end

# An element of a row of an adjacency list of a weighted undirected graph.
Neighbor = Struct.new(:dest, :wt)

def read_record
  gets.split.map!(&:to_i)
end

FIRST_VERTEX = 1 # The lowest vertex number used.

UNREACHABLE_COST = -1 # The fictitious cost displayed for unreachable vertices.

# Creates an adjacency list with all rows empty.
def empty_adjacency_list(vertex_count)
  raise ArgumentError, 'negative vertex count?' if vertex_count.negative?

  adj = Array.new(vertex_count) { [] }
  FIRST_VERTEX.times { adj.unshift(nil) }
  adj
end

# Reads a graph as a hash-based sparse adjacency matrix.
def read_simple_graph_edges(edge_count)
  raise ArgumentError, 'negative edge count?' if edge_count.negative?

  edges = {}

  edge_count.times do
    u, v, wt = read_record
    next if u == v

    uv = Endpoints.new(u, v)
    old_wt = edges[uv]
    edges[uv] = wt if old_wt.nil? || wt < old_wt
  end

  edges
end

# Reads a weighted undirected graph as an adjacency list.
def read_graph
  vertex_count, edge_count = read_record

  adj = empty_adjacency_list(vertex_count)

  read_simple_graph_edges(edge_count).each do |uv, wt|
    u, v = *uv
    adj[u] << Neighbor.new(v, wt)
    adj[v] << Neighbor.new(u, wt)
  end

  adj
end

# Calculates the least cost paths from a given starting vertex, start, to each
# vertex in a graph. Uses Dijkstra's algorithm.
def compute_shortest_reaches(adj, start)
  costs = Array.new(adj.size, UNREACHABLE_COST)

  cost_heap = PrimHeap.new(FIRST_VERTEX, adj.size - 1)
  cost_heap.decrease(start, 0)

  until cost_heap.empty?
    src, src_cost = *cost_heap.pop
    break if src_cost == Float::INFINITY

    costs[src] = src_cost

    adj[src].each do |neighbor|
      cost_heap.decrease(neighbor.dest, src_cost + neighbor.wt)
    end
  end

  costs
end

gets.to_i.times do
  adj = read_graph
  start = gets.to_i
  costs = compute_shortest_reaches(adj, start)
  puts((costs[FIRST_VERTEX...start] + costs[(start + 1)..-1]).join(' '))
end
